### PR TITLE
Feat/model extrusion

### PR DESCRIPTION
**Pull request by francoisbonneau**

### CODE REVIEW

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 1]
```diff
@@ -59,10 +68,310 @@ namespace
...
     }
+
+    class SectionExtruder
+    {
+    public:
+        SectionExtruder( const geode::Section& section,
+            geode::index_t axis_to_extrude,
+            double min_coordinate,
+            double max_coordinate )
+            : section_{ section },
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:84]: const & cannot be initialized using { }, this is a bug in old gcc

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 2]
```diff
@@ -59,10 +68,310 @@ namespace
...
+            section_slice0_mapping_ = std::get< 1 >( slice0_conversion_output );
+            slice0_brep_mapping_ =
+                brep_builder_.copy( std::get< 0 >( slice0_conversion_output ) );
+
+            auto slice1_conversion_output = convert_section_into_brep(
+                section_, axis_to_extrude, max_coordinate );
+            section_slice1_mapping_ = std::get< 1 >( slice1_conversion_output );
+            geode::BRepConcatener brep_concatener( brep_ );
+            slice1_brep_mapping_ = brep_concatener.concatenate(
+                std::get< 0 >( slice1_conversion_output ) );
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:103]: It looks like this is part of the process, why is it in the constructor?

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 3]
```diff
@@ -59,10 +68,310 @@ namespace
...
+        }
+
+        ~SectionExtruder() = default;
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:106]:
```suggestion
```

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 4]
```diff
@@ -59,10 +68,310 @@ namespace
...
+
+    private:
+        geode::uuid section_slice0_brep_mapping(
+            const geode::ComponentType& brep_component_type,
+            geode::uuid section_component )
+        {
+            return (
+                slice0_brep_mapping_.at( brep_component_type )
+                    .in2out( section_slice0_mapping_.at( brep_component_type )
+                                 .in2out( section_component ) ) );
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:123]:
```suggestion
            return 
                slice0_brep_mapping_.at( brep_component_type )
                    .in2out( section_slice0_mapping_.at( brep_component_type )
                                 .in2out( section_component ) );
```
This is not needed and verbose

Comment (BotellaA) [reply src/geode/model/helpers/convert_brep_section.cpp:123]: Replace for all the file

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 5]
```diff
@@ -59,10 +68,310 @@ namespace
...
+        }
+
+        void create_lines()
+        {
+            for( const auto& corner : section_.corners() )
+            {
+                extrude_corner( corner );
+            }
+        }
+        geode::uuid extrude_corner( const geode::Corner2D& section_corner )
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:142]:
```suggestion

        geode::uuid extrude_corner( const geode::Corner2D& section_corner )
```
Add spacing before and after a method

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 6]
```diff
@@ -59,10 +68,310 @@ namespace
...
+            }
+        }
+        geode::uuid extrude_line( const geode::Line2D& section_line )
+        {
+            const auto& line_slice0 = brep_.line( section_slice0_brep_mapping(
+                geode::Line3D::component_type_static(), section_line.id() ) );
+            const auto& line_slice1 = brep_.line( section_slice1_brep_mapping(
+                geode::Line3D::component_type_static(), section_line.id() ) );
+            const auto& surface = brep_.surface( brep_builder_.add_surface(
+                geode::OpenGeodePolygonalSurface< 3 >::impl_name_static() ) );
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:185]:
```suggestion
                 ) );
```
Default is polygonal.

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 7]
```diff
@@ -59,10 +68,310 @@ namespace
...
+            brep_builder_.add_line_surface_boundary_relationship(
+                line_slice0, surface );
+            brep_builder_.add_line_surface_boundary_relationship(
+                line_slice1, surface );
+            add_extruded_line_surface_boundary_relationship(
+                section_line, surface );
+            auto surface_builder =
+                brep_builder_.surface_mesh_builder( surface.id() );
+            auto prev_min_pt =
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:194]:
```suggestion
            const auto prev_min_pt =
```

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 8]
```diff
@@ -59,10 +68,310 @@ namespace
...
+                surface_builder->create_point( line_slice0.mesh().point( 0 ) );
+            brep_builder_.set_unique_vertex(
+                { surface.component_id(), prev_min_pt },
+                brep_.unique_vertex( { line_slice0.component_id(), 0 } ) );
+            auto prev_max_pt =
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:199]:
```suggestion
            const auto prev_max_pt =
```

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 9]
```diff
@@ -59,10 +68,310 @@ namespace
...
+                surface_builder->create_point( line_slice1.mesh().point( 0 ) );
+            brep_builder_.set_unique_vertex(
+                { surface.component_id(), prev_max_pt },
+                brep_.unique_vertex( { line_slice1.component_id(), 0 } ) );
+            for( const auto v : geode::Range( line_slice0.mesh().nb_edges() ) )
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:204]:
```suggestion
            for( const auto v : geode::Range{ line_slice0.mesh().nb_edges() } )
```

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 10]
```diff
@@ -59,10 +68,310 @@ namespace
...
+            {
+                const auto min_pt = surface_builder->create_point(
+                    line_slice0.mesh().point( v + 1 ) );
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:207]: This is not always true. Vertices in an EdgedCurve are not ordered.

**Review by BotellaA — CHANGES_REQUESTED** [review]

File: bindings/python/tests/model/test-py-convert-brep.py [hunk 11]
```diff
@@ -43,3 +38,32 @@
...
+    brep = model.extrude_section_to_brep(section, 2, 0., 10.)
+
+    if brep.nb_corners() != 2*section.nb_corners():
+        raise ValueError("[Test] extruded BRep have wrong number of corners")
+    if brep.nb_lines() != 2*section.nb_lines()+section.nb_corners():
+        raise ValueError("[Test] extruded BRep have wrong number of lines")
+    if brep.nb_surfaces() != 2*section.nb_surfaces()+section.nb_lines():
+        raise ValueError("[Test] extruded BRep have wrong number of surfaces")
+    if brep.nb_blocks() != section.nb_surfaces():
+        raise ValueError("[Test] extruded BRep have wrong number of blocks")
```
Comment (panquez) [inline bindings/python/tests/model/test-py-convert-brep.py:61]:
```suggestion
    if brep.nb_corners() != 2*section.nb_corners():
        raise ValueError("[Test] Extruded BRep have wrong number of corners")
    if brep.nb_lines() != 2*section.nb_lines()+section.nb_corners():
        raise ValueError("[Test] Extruded BRep have wrong number of lines")
    if brep.nb_surfaces() != 2*section.nb_surfaces()+section.nb_lines():
        raise ValueError("[Test] Extruded BRep have wrong number of surfaces")
    if brep.nb_blocks() != section.nb_surfaces():
        raise ValueError("[Test] Extruded BRep have wrong number of blocks")
```

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 12]
```diff
@@ -59,10 +68,310 @@ namespace
...
+            auto slice0_conversion_output = convert_section_into_brep(
+                section_, axis_to_extrude, min_coordinate );
+            section_slice0_mapping_ = std::get< 1 >( slice0_conversion_output );
+            slice0_brep_mapping_ =
+                brep_builder_.copy( std::get< 0 >( slice0_conversion_output ) );
+
+            auto slice1_conversion_output = convert_section_into_brep(
+                section_, axis_to_extrude, max_coordinate );
+            section_slice1_mapping_ = std::get< 1 >( slice1_conversion_output );
+            geode::BRepConcatener brep_concatener( brep_ );
```
Comment (panquez) [inline src/geode/model/helpers/convert_brep_section.cpp:101]:
```suggestion
            geode::BRepConcatener brep_concatener{ brep_ };
```

File: src/geode/model/helpers/convert_brep_section.cpp (hunk 5 above)
Comment (panquez) [inline src/geode/model/helpers/convert_brep_section.cpp:142]: You do not use the returned uuid. Is it intentional?

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 13]
```diff
@@ -135,4 +444,17 @@ namespace geode
         copy_unique_vertices( section, builder, mappings );
         return std::make_tuple( std::move( brep ), std::move( mappings ) );
     }
+
+    BRep extrude_section_to_brep( const Section& section,
+        index_t axis_to_extrude,
+        double min_coordinate,
+        double max_coordinate )
+    {
+        OPENGEODE_ASSERT( min_coordinate < max_coordinate,
```
Comment (panquez) [inline src/geode/model/helpers/convert_brep_section.cpp:453]: ASSERT or EXCEPTION? Is it really needed?

**Review by panquez — APPROVED** [review]

File: src/geode/model/helpers/convert_brep_section.cpp [hunk 14]
```diff
@@ -59,10 +68,325 @@ namespace
...
+                    { line_slice0.component_id(), prev_slide0_pointid } ) );
+            const auto prev_slide1_pointid =
+                line_slice1.mesh().edge_vertex( { 0, 0 } );
+            auto prev_slice1_pt = surface_builder->create_point(
+                line_slice1.mesh().point( prev_slide1_pointid ) );
+            brep_builder_.set_unique_vertex(
+                { surface.component_id(), prev_slice1_pt },
+                brep_.unique_vertex(
+                    { line_slice1.component_id(), prev_slide1_pointid } ) );
+            for( const auto v : geode::Range{ line_slice0.mesh().nb_edges() } )
```
Comment (BotellaA) [inline src/geode/model/helpers/convert_brep_section.cpp:211]: Maybe v is not a good iterator name since it is more edge than vertex related

**Review by BotellaA — APPROVED** [review]

**Pull request merged**